hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ProcessorTasklet.java

```diff
@@ -314,47 +290,46 @@ private void stateMachineStep() {
                 if (outbox.offerToEdgesAndSnapshot(currentBarrier)) {
                     progTracker.madeProgress();
                     if (currentBarrier.isTerminal()) {
-                        state = EMIT_DONE_ITEM;
+                        state = WAITING_FOR_SNAPSHOT_COMPLETED;
                     } else {
                         currentBarrier = null;
                         receivedBarriers.clear();
-                        pendingSnapshotId++;
-                        state = initialProcessingState();
+                        pendingSnapshotId1++;
+                        state = processingState();
                     }
                 }
-                progTracker.notDone();
                 return;
 
-            case COMPLETE:
-                progTracker.notDone();
-                // check ssContext to see if a barrier should be emitted
-                long currSnapshotId = ssContext.activeSnapshotId();
-                assert currSnapshotId + 1 == pendingSnapshotId || currSnapshotId == pendingSnapshotId
-                        : "Unexpected new snapshot id: " + currSnapshotId + ", expected was "
-                                + (pendingSnapshotId - 1) + " or " + pendingSnapshotId;
-                if (currSnapshotId == pendingSnapshotId) {
-                    if (outbox.hasUnfinishedItem()) {
-                        outbox.block();
-                    } else {
-                        outbox.unblock();
-                        state = SAVE_SNAPSHOT;
-                        currentBarrier = new SnapshotBarrier(currSnapshotId, ssContext.isTerminalSnapshot());
-                        progTracker.madeProgress();
-                        return;
-                    }
-                }
-                if (processor.complete()) {
+            case ON_SNAPSHOT_COMPLETED:
+            case ON_SNAPSHOT_COMPLETED_BEFORE_END:
+                if (processor.onSnapshotCompleted(ssContext.isLastPhase1Successful())) {
+                    pendingSnapshotId2++;
+                    ssContext.phase2DoneForTasklet();
                     progTracker.madeProgress();
-                    state = EMIT_DONE_ITEM;
+                    state = state == ON_SNAPSHOT_COMPLETED_BEFORE_END
+                            ? EMIT_DONE_ITEM
+                            : processingState();
+                }
+                return;
+
+            case WAITING_FOR_SNAPSHOT_COMPLETED:
+                long currSnapshotId2 = ssContext.activeSnapshotIdPhase2();
+                if (currSnapshotId2 >= pendingSnapshotId2) {
+                    state = ON_SNAPSHOT_COMPLETED_BEFORE_END;
+                    stateMachineStep(); // recursion
                 }
                 return;
 
+            case COMPLETE:
+                complete();
```

Comment:
We should probably have each clause in the switch statement at the same semantic level. Either inline everything in each, or methods in each. Basically same as my previous comment about refactoring the switch.